test/lib/active_record/connection_adapters/trilogy_adapter_test.rb

```diff
@@ -27,6 +27,32 @@ class ActiveRecord::ConnectionAdapters::TrilogyAdapterTest < TestCase
     @adapter.disconnect!
   end
 
+  test ".new_client" do
+    client = @adapter.class.new_client(@configuration)
+    assert_equal Trilogy, client.class
+  end
+
+  test ".new_client on db error" do
+    configuration = @configuration.merge(database: "unknown")
+    assert_raises ActiveRecord::NoDatabaseError do
+      @adapter.class.new_client(configuration)
+    end
+  end
+
+  test ".new_client on access denied error" do
+    configuration = @configuration.merge(username: "unknown")
+    assert_raises ActiveRecord::DatabaseConnectionError do
+      @adapter.class.new_client(configuration)
+    end
+  end
+
+  test ".new_client on host error" do
+    configuration = @configuration.merge(host: "unknown")
+    assert_raises ActiveRecord::DatabaseConnectionError do
```

Comment:
@composerinteralia this fails right now, because we raise this as a query error so it's not being rescued and translated.

```  1) Failure:
ActiveRecord::ConnectionAdapters::TrilogyAdapterTest#test_.new_client_on_host_error [/Users/adriannachang/src/github.com/activerecord-trilogy-adapter/test/lib/active_record/connection_adapters/trilogy_adapter_test.rb:51]:
[ActiveRecord::DatabaseConnectionError] exception expected, not
Class: <Trilogy::QueryError>
Message: <"trilogy_connect - unable to connect to unknown:3306: TRILOGY_DNS_ERR">
```

But on main currently, this type of error shows up as a `Trilogy::Error` and not a `Trilogy::DatabaseError`, so I don't believe it's being rescued properly right now either 🤔 :

```
ActiveRecord::ConnectionAdapters::TrilogyAdapterTest#test_.new_client_on_host_error [/Users/adriannachang/src/github.com/activerecord-trilogy-adapter/test/lib/active_record/connection_adapters/trilogy_adapter_test.rb:32]:
[ActiveRecord::NoDatabaseError] exception expected, not
Class: <Trilogy::Error>
Message: <"trilogy_connect - unable to connect to unknown:3306: TRILOGY_DNS_ERR">
```

This ought to be raised as a `BaseConnectionError` right?

Reply:
Yeah, I agree QueryError doesn't quite make sense for this one. We could add a case for that one to handle_trilogy_error. This does make me wonder if there are other [trilogy errors](https://github.com/github/trilogy/blob/main/inc/trilogy/error.h#L7-L24) that are going to turn up as `QueryError`s but maybe make more sense as something else.

Reply:
> BaseConnectionError is for "potentially transient network errors", correct? Does that fit this DNS error?

I think so 👍 It's the network refusing the connection, and [Mysql2 treats these as `ConnectionError`s](https://www.rubydoc.info/gems/mysql2/Mysql2/Error#ConnectionError-constant) so it seems reasonable for us to do the same.

Reply:
Okay, so I started looking at this in the Trilogy client, but I'm a little stuck on which layer to change the error 😅 We could change the [`#initialize` logic](https://github.com/github/trilogy/blob/main/contrib/ruby/ext/trilogy-ruby/cext.c#L487-L494) to raise `BaseConnectionError` directly instead of calling `#handle_trilogy_error`, but there's no relevant error code to set... So I'm wondering if this is a problem for the Trilogy client code directly rather than the bindings -- perhaps we need to raise something other than `TRILOGY_DNS_ERR`? 

Either way, perhaps it makes sense to skip the host test for now so that we can get this initial work shipped, and then I'll investigate what we should be doing in a follow up. WDYT?

Reply:
Jean and I chatted IRL -- seems like the most appropriate solution is just to add all of the relevant error code ints to `#handle_trilogy_error` and have those raised as connection errors. `TRILOGY_DNS_ERR` for one, but also `TRILOGY_EOF` and `TRILOGY_PROTOCOL_VIOLATION` (as evidenced in the [connection tests that fails with MySQL8 right now](https://github.com/github/trilogy/actions/runs/3911145536/jobs/6684193014))